src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,313 @@
+        var acknowledgementProcessorTask = ProcessAcknowledgementChannelMessagesAsync(cancellationToken);
+        var topicMessageProcessorTask = ProcessTopicChannelMessagesAsync(cancellationToken);
+
+        try
+        {
+            await Task.WhenAll(fetchMessagesTask, acknowledgementProcessorTask, topicMessageProcessorTask);
+        }
+        catch (OperationCanceledException)
+        {
+            // Will be cleaned up during DisposeAsync
+        }
+    }
+
+    /// <summary>
+    /// Retrieves or creates the bidirectional stream to the DaprClient for transacting pub/sub subscriptions.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>The stream connection.</returns>
+    private async Task<AsyncDuplexStreamingCall<P.SubscribeTopicEventsRequestAlpha1, P.SubscribeTopicEventsResponseAlpha1>> GetStreamAsync(CancellationToken cancellationToken)
+    {
+        await semaphore.WaitAsync(cancellationToken);
+
+        try
+        {
+            return clientStream ??= client.SubscribeTopicEventsAlpha1(cancellationToken: cancellationToken);
+        }
+        finally
+        {
+            semaphore.Release();
+        }
+    }
+
+    /// <summary>
+    /// Acknowledges the indicated message back to the Dapr sidecar with an indicated behavior to take on the message.
+    /// </summary>
+    /// <param name="messageId">The identifier of the message the behavior is in reference to.</param>
+    /// <param name="behavior">The behavior to take on the message as indicated by either the message handler or timeout message handling configuration.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns></returns>
+    private async Task AcknowledgeMessageAsync(string messageId, TopicResponseAction behavior, CancellationToken cancellationToken)
+    {
+        var action = behavior switch
+        {
+            TopicResponseAction.Success => TopicEventResponse.Types.TopicEventResponseStatus.Success,
+            TopicResponseAction.Retry => TopicEventResponse.Types.TopicEventResponseStatus.Retry,
+            TopicResponseAction.Drop => TopicEventResponse.Types.TopicEventResponseStatus.Drop,
+            _ => throw new InvalidOperationException(
+                $"Unrecognized topic acknowledgement action: {behavior}")
+        };
+
+        var acknowledgement = new TopicAcknowledgement(messageId, action);
+        await acknowledgementsChannel.Writer.WriteAsync(acknowledgement, cancellationToken);
+    }
+
+    /// <summary>
+    /// Processes each acknowledgement from the acknowledgement channel reader as it's populated.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    private async Task ProcessAcknowledgementChannelMessagesAsync(CancellationToken cancellationToken)
+    {
+        var messageStream = await GetStreamAsync(cancellationToken);
```

Comment:
The stream has already been acquired when this method is called, so it can just be passed to it.  In that case, there is (i.e. should be) no need for the semaphore.

Reply:
Good catch